pallets/regulated-assets/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+        false
+    }
+
+    /// Mapping from asset id to whether it is regulated or not
+    #[pallet::storage]
+    #[pallet::getter(fn regulated_asset)]
+    pub type RegulatedAsset<T: Config> =
+        StorageMap<_, Identity, AssetIdOf<T>, bool, ValueQuery, DefaultRegulatedAsset<T>>;
+
+    /// Mapping from SBT (asset_id) to its metadata
+    #[pallet::storage]
+    #[pallet::getter(fn soulbound_asset)]
+    pub type SoulboundAsset<T: Config> =
+        StorageMap<_, Identity, AssetIdOf<T>, SoulboundTokenMetadata<AssetIdOf<T>>, OptionQuery>;
+
+    /// Mapping from `asset_id` to its SBTs which grant permission to transfer, mint, and burn the `asset_id`
+    #[pallet::storage]
+    #[pallet::getter(fn sbts_by_asset)]
+    pub type SBTsByAsset<T: Config> =
+        StorageMap<_, Identity, AssetIdOf<T>, BTreeSet<AssetIdOf<T>>, ValueQuery>;
+}
+
+impl<T: Config> AssetRegulator<AccountIdOf<T>, AssetIdOf<T>> for Pallet<T> {
+    fn check_permission(
+        issuer: &AccountIdOf<T>,
+        affected_account: &AccountIdOf<T>,
+        asset_id: &AssetIdOf<T>,
+        permission_id: &PermissionId,
+    ) -> Result<(), DispatchError> {
+        if Self::soulbound_asset(asset_id).is_some() {
+            // Check if the issuer is the asset owner
+            let is_asset_owner = <T as Config>::AssetInfoProvider::is_asset_owner(asset_id, issuer);
+
+            if is_asset_owner {
+                // Asset owner of the SBT can do all asset operations except transfer
+                if permission_id == &TRANSFER {
+                    return Err(Error::<T>::SoulboundAssetNotTransferable.into());
+                }
+                return Ok(());
+            } else {
+                return Err(Error::<T>::SoulboundAssetNotOperationable.into());
+            }
+        }
+
+        // If asset is not regulated, then no need to check permissions
+        if !Self::regulated_asset(asset_id) {
+            return Ok(());
+        }
+
+        // If the account is a technical account, then it can do all operations
+        if Technical::<T>::lookup_tech_account_id(issuer).is_ok() {
+            return Ok(());
+        }
+
+        let sbts = Self::sbts_by_asset(asset_id);
+
+        let issuer_has_sbt = sbts.iter().any(|sbt| {
+            <T as Config>::AssetInfoProvider::total_balance(sbt, issuer)
+                .map_or(false, |balance| balance > 0)
+        });
+
+        let affected_account_has_sbt = sbts.iter().any(|sbt| {
+            <T as Config>::AssetInfoProvider::total_balance(sbt, affected_account)
+                .map_or(false, |balance| balance > 0)
+        });
+
+        if !issuer_has_sbt || !affected_account_has_sbt {
+            return Err(Error::<T>::AllInvolvedUsersShouldHoldSBT.into());
+        }
+
+        Ok(())
+    }
+
+    fn assign_permission(
+        _owner: &AccountIdOf<T>,
+        _asset_id: &AssetIdOf<T>,
+        _permission_id: &PermissionId,
+    ) -> Result<(), DispatchError> {
+        Ok(())
```

Comment:
@Alexey-N-Chernyshov There is no need for logic to be implemented here so far. 
It should return Ok because of its implementation for [(A, B) ](https://github.com/sora-xor/sora2-network/blob/1012-tech-pallet-for-regulated-assets-support/common/src/traits.rs#L1469)